File: scala-advent-of-code/src/main/scala/advent/twenty_sixteen/Day03SquaresWithThreeSides.scala
package advent.twenty_sixteen

import advent.common.DailyProblem
import advent.utilities.FileUtilities

class Day03SquaresWithThreeSides(filename: String) extends DailyProblem[Int, Int] {

  case class Triangle(a: Int, b: Int, c: Int) {
    def isValid(): Boolean =
      if (a > b && a > c) a < b + c //
      else if (b > a && b > c) b < a + c //
      else c < a + b
  }

  private val lines = FileUtilities.readFile(filename)

  private val linePattern = "\\s*([0-9]+)\\s*([0-9]+)\\s*([0-9]+)".r

  private val triangleData = lines.map(line => line match {
    case linePattern(a, b, c) => (a.toInt, b.toInt, c.toInt)
  })

  private val possibleTrianglesPart1 = triangleData.map(row => Triangle(row._1, row._2, row._3))

  private val possibleTrianglesPart2 = for {i <- 0.to(triangleData.size - 1).by(3)
                                            row1 = triangleData(i)
                                            row2 = triangleData(i + 1)
                                            row3 = triangleData(i + 2)}
    yield List(Triangle(row1._1, row2._1, row3._1), Triangle(row1._2, row2._2, row3._2), Triangle(row1._3, row2._3, row3._3))

  override val part1Answer: Int = possibleTrianglesPart1.filter(t => t.isValid()).size
  override val part2Answer: Int = possibleTrianglesPart2.flatten.filter(t => t.isValid()).size
}

File: java-advent-of-code/src/main/java/advent/twenty_sixteen/Day03SquaresWithThreeSides.java
package advent.twenty_sixteen;

import advent.common.DailyProblem;
import advent.common.Triple;
import advent.utilities.FileUtilities;
import advent.utilities.Parsers;

import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

public class Day03SquaresWithThreeSides implements DailyProblem<Integer, Integer> {
    private static final Function<String, Triple<Integer, Integer, Integer>> TO_TRIPLE_INT = line -> {
        String[] bits = line.trim().split("\\s+");
        return new Triple<>(Integer.parseInt(bits[0]), Integer.parseInt(bits[1]), Integer.parseInt(bits[2]));
    };

    private final int part1Answer;
    private final int part2Answer;

    public Day03SquaresWithThreeSides(String filename) {
        List<String> input = FileUtilities.readLines(filename, Parsers::TO_STRING);

        List<Triple<Integer, Integer, Integer>> digits = input.stream().map(TO_TRIPLE_INT).collect(Collectors.toList());

        this.part1Answer = (int) digits.stream().filter(t -> isValid(t)).count();

        int count = 0;
        for (int i = 0; i < digits.size(); i += 3) {
            if (isValid(digits.get(i).getFirst(), digits.get(i + 1).getFirst(), digits.get(i + 2).getFirst()))
                count++;
            if (isValid(digits.get(i).getSecond(), digits.get(i + 1).getSecond(), digits.get(i + 2).getSecond()))
                count++;
            if (isValid(digits.get(i).getThird(), digits.get(i + 1).getThird(), digits.get(i + 2).getThird()))
                count++;
        }

        this.part2Answer = count;
    }

    private boolean isValid(Triple<Integer, Integer, Integer> triple) {
        int a = triple.getFirst();
        int b = triple.getSecond();
        int c = triple.getThird();
        return isValid(a, b, c);
    }

    private boolean isValid(int a, int b, int c) {
        if (a > b && a > c) {
            return a < b + c;
        }
        if (b > a && b > c) {
            return b < a + c;
        }
        return c < a + b;
    }

    @Override
    public Integer getPart1Answer() {
        return part1Answer;
    }

    @Override
    public Integer getPart2Answer() {
        return part2Answer;
    }
}
